storage/innobase/dict/dict0crea.cc

```diff
@@ -1865,29 +1831,8 @@ dict_create_add_foreign_to_dictionary(
 				      , name, foreign->id, trx);
 
 	if (error != DB_SUCCESS) {
-
-		if (error == DB_DUPLICATE_KEY) {
-			char	buf[MAX_TABLE_NAME_LEN + 1] = "";
-			char	tablename[MAX_TABLE_NAME_LEN + 1] = "";
-			char*	fk_def;
-
-			innobase_convert_name(tablename, MAX_TABLE_NAME_LEN,
-				name, strlen(name), trx->mysql_thd);
-
-			innobase_convert_name(buf, MAX_TABLE_NAME_LEN,
-				foreign->id, strlen(foreign->id), trx->mysql_thd);
-
-			fk_def = dict_foreign_def_get((dict_foreign_t*)foreign, trx);
-
-			ib_push_warning(trx, error,
-				"Create or Alter table %s with foreign key constraint"
-				" failed. Foreign key constraint %s"
-				" already exists on data dictionary."
-				" Foreign key constraint names need to be unique in database."
-				" Error in foreign key definition: %s.",
-				tablename, buf, fk_def);
```

Comment:
Can we preserver the important information in this warning (the same is reflected in the test result) along with the general error message in innodb_fk_error ? I think it is ok to have warnings pushed from each specific path with more details.

Reply:
This appears to be the only path that can actually output the foreign key error. The other call should always have crashed, because we are passing an integer to a `"%s"` format parameter.

But, I agree that it would be useful to somehow mention that the constraint name is already being used in the current schema.